### PR TITLE
VCL variables

### PR DESCRIPTION
**Pull request by szymi-**

### CODE REVIEW

File: vaas-app/src/vaas/cluster/models.py [hunk 1]
```diff
@@ -53,6 +56,9 @@ def __str__(self):
     def get_template_version(self):
         return self.version
 
+    def clean(self):
+        vcl_variable_validator(self.content, self.pk, VclVariable, VarnishServer)
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/models.py]: is this correct place for injecting validators ?

File: vaas-app/src/vaas/manager/signals.py [hunk 2]
```diff
@@ -162,6 +169,51 @@ def clean_up_tags(sender, **kwargs):
...
+
+    if sender.__name__ not in settings.REFRESH_TRIGGERS_CLASS:
+        return
+
+    if settings.SIGNALS != 'on':
+        return
+
+    instance = kwargs['instance']
+    logger.debug("sender: " + str(sender))
+    logger.debug("instance: " + str(instance))
```
Comment (ziollek) [inline vaas-app/src/vaas/manager/signals.py]: this part can be shared between pre_ & post_ signal hooks

File: vaas-app/src/vaas/cluster/models.py [hunk 3]
```diff
@@ -1,13 +1,16 @@
 # -*- coding: utf-8 -*-
 
+import re
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/models.py]: unused import?

File: vaas-app/src/vaas/settings/base.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 import os
 import environ
+import re
```
Comment (ziollek) [inline vaas-app/src/vaas/settings/base.py]: unused import ?

File: vaas-app/src/vaas/validators.py [hunk 5]
```diff
@@ -2,7 +2,26 @@
...
+
+class VclVariableValidator(ValidationError):
+    pass
+
+
+def vcl_variable_validator(vcl_content, vcl_pk, vcl_variable_class, varnish_server_class):
+
+    pattern = re.compile('#{.*}', re.MULTILINE)
+    placeholders = re.findall(pattern, vcl_content)
+    placeholders = [p[2:-1] for p in placeholders]
```
Comment (ziollek) [inline vaas-app/src/vaas/validators.py]: nonoptimal data structure - placeholders can be repeated in list - you can use set

>>> pattern = re.compile('a')
>>> string = 'abceabcdabcd'
>>> re.findall(pattern, string)
['a', 'a', 'a']

File: vaas-app/src/vaas/vcl/renderer.py [hunk 6]
```diff
@@ -205,6 +220,7 @@ def __init__(self):
         self.directors.sort(key=lambda director: ROUTE_SETTINGS[director.router]['priority'])
         self.dcs = list(Dc.objects.all())
         self.template_blocks = list(VclTemplateBlock.objects.all().prefetch_related('template'))
+        self.vcl_variables = list(VclVariable.objects.all().prefetch_related('cluster'))
```
Comment (ziollek) [inline vaas-app/src/vaas/vcl/renderer.py]: unused variable ?

File: vaas-app/src/vaas/settings/base.py (hunk 4 above)
Comment (TheMickeyMike) [inline vaas-app/src/vaas/settings/base.py]: This is unused import

File: vaas-app/src/vaas/vcl/tests/expected-vcl-3.0.vcl [hunk 7]
```diff
@@ -356,4 +356,5 @@ sub vcl_recv {
     }
     return (lookup);
 }
-## other functions ##
\ No newline at end of file
+## other functions ##
+#vcl_variable_content
```
Comment (ziollek) [inline vaas-app/src/vaas/vcl/tests/expected-vcl-3.0.vcl]: convention ## comment ## ?

File: vaas-app/src/vaas/validators.py [hunk 8]
```diff
@@ -2,7 +2,26 @@
...
 
 vcl_name_re = re.compile(r'^[a-zA-Z0-9_]+$')
 vcl_name_validator = RegexValidator(vcl_name_re, "Allowed characters: letters, numbers and underscores.", 'invalid')
+
+
+class VclVariableValidator(ValidationError):
+    pass
+
+
+def vcl_variable_validator(vcl_content, vcl_pk, vcl_variable_class, varnish_server_class):
```
Comment (ziollek) [inline vaas-app/src/vaas/validators.py]: please add unit tests

File: vaas-app/src/vaas/cluster/models.py (hunk 3 above)
Comment (TheMickeyMike) [inline vaas-app/src/vaas/cluster/models.py]: Unuseed import

File: vaas-app/src/vaas/cluster/models.py (hunk 1 above)
Comment (szymi-) [reply vaas-app/src/vaas/cluster/models.py]: According to Django docs: "This method should be used to provide custom model validation" .

Comment (awmackowiak) [reply vaas-app/src/vaas/cluster/models.py]: HINT: a model’s clean() method is not invoked when you call your model’s save() method.

**Pull request merged**